route-rs-packets/src/arp.rs

```diff
@@ -0,0 +1,253 @@
+use crate::{EthernetFrame, MacAddr};
+use std::convert::TryInto;
+use std::net::IpAddr;
+
+pub enum ArpOp {
+    Request = 1,
+    Reply = 2,
+}
+
+pub enum ArpHardwareType {
+    Ethernet = 1,
+}
+
+// TODO: should move to types.rs or ethernet.rs
+pub const IPV4_ETHER_TYPE: u16 = 0x0800;
+pub const IPV6_ETHER_TYPE: u16 = 0x86DD;
+pub const ARP_ETHER_TYPE: u16 = 0x0806;
+
+const HARDWARE_TYPE_RANGE: (usize, usize) = (0, 2);
+const PROTOCOL_TYPE_RANGE: (usize, usize) = (2, 4);
+const HARDWARE_ADDR_LEN_RANGE: (usize, usize) = (4, 5);
+const PROTOCOL_ADDR_LEN_RANGE: (usize, usize) = (5, 6);
+const OPCODE_RANGE: (usize, usize) = (6, 8);
+
+///
+/// EthernetFrame wrapper with getters/setters for the packet structure described in RFC 826
+/// https://tools.ietf.org/html/rfc826
+///
+// NOTE: Could be implemented in various ways, such as a specialized version of EthernetFrame that's
+// known to be an ARP frame. It could be implemented in a similar way that packets are promoted/demoted
+// with `TryFrom`.
+#[derive(Clone)]
+pub struct ArpFrame {
+    frame: EthernetFrame,
+}
+
+impl ArpFrame {
+    // TODO: would like default args for Ethernet (6 bytes) and IPv4 (4 bytes)
```

Comment:
I need to set an empty ARP frame to have _some_ values here, since the payload is variable sized. We run into some subtle bugs if we don't set these values to something reasonable, e.g. If we attempt to set any of the variable length payload fields while these lengths are 0, we'll have nowhere to write to.

Since we're wrapping `EthernetFrame` here, I think it's ok to provide a few more guardrails.

Reply:
Typically I've handled Length fields by setting them appropriately compared to the length of the buffer we make.  I've also noticed the protocols usually specify a valid minimum value.  Does something like that apply here?

Reply:
I guess the short answer here, is pick some reasonable defaults, and document it in the doc comments.

Reply:
Maybe? I could try to find the minimum hardware/protocol addr lengths that ARP supports, but those parameters would only work with that particular hardware/protocol pair. I think I'll brainstorm some better constructor ideas to make sure the packet gets initialized into a consistent state.